### PR TITLE
add ad4170 device driver

### PR DESCRIPTION
**Pull request by buha**
Initial driver version with a device driver API that covers direct register access and higher level chip specific functionality.

## Pull Request Description

Please replace this with a detailed description and motivation of the changes. 
You can tick the checkboxes below with an 'x' between square brackets or just check them after publishing the PR. 
If this PR contains a breaking change, list dependent PRs and try to push all related PRs at the same time.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

### CODE REVIEW

**Comment by jansunil** [comment]
Hi  @buha - The same copy of the driver has been tested in the IIO FW. I confirm that this needs no mode changes and the version is stable for a release.

**Comment by buha** [comment]
Cool, thanks for confirming @jansunil !

I pushed a new version which shows @jansunil and @mphalke as authors in the .c file, they contributed as well.

**Comment by buha** [comment]
V2:
- fix attempt for doxygen errors
- renamed active_dev to id

**Comment by buha** [comment]
CI fails building xilinx projects, it's unrelated.

File: drivers/adc/ad4170/ad4170.h [hunk 1]
```diff
@@ -0,0 +1,1142 @@
...
+	/** DIG_AUX2 GPIO descriptor. */
+	struct no_os_gpio_desc *gpio_dig_aux2;
+	/** Specifies whether the SPI+GPIO is interface is initialized. */
+	bool digif;
+};
+
+/** ad4170_config default state when the chip is hardware/software reset. */
+extern struct ad4170_config ad4170_config_reset;
+
+int32_t ad4170_spi_reg_read(struct ad4170_dev *dev,
```
Comment (amiclaus) [inline drivers/adc/ad4170/ad4170.h:1066]: we should stick with our convention of using `int` type for functions.

File: drivers/adc/ad4170/ad4170.c [hunk 2]
```diff
@@ -0,0 +1,2123 @@
...
+	return _ad4170_read(dev, pbuf, nb_samples, AD4170_REG_DATA_32b);
+}
+
+static int32_t _scratchpad_test(struct ad4170_dev *dev, uint8_t pattern)
+{
+	int32_t ret;
+	uint32_t reg;
+
+	ret = ad4170_spi_reg_write(dev, AD4170_REG_SCRATCH_PAD, pattern);
+	if (ret < 0)
```
Comment (amiclaus) [inline drivers/adc/ad4170/ad4170.c:737]: same goes for `if (ret)` convention linux compliant.

File: drivers/adc/ad4170/ad4170.c [hunk 3]
```diff
@@ -0,0 +1,2123 @@
...
+#include <stdbool.h>
+#include <string.h>
+#include "no_os_error.h"
+#include "no_os_delay.h"
+#include "no_os_crc8.h"
+#include "ad4170.h"
+
+NO_OS_DECLARE_CRC8_TABLE(ad4170_crc8);
+
+struct ad4170_config ad4170_config_reset = {
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:52]: `static`

File: drivers/adc/ad4170/ad4170.c [hunk 4]
```diff
@@ -0,0 +1,2123 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include "no_os_error.h"
+#include "no_os_delay.h"
+#include "no_os_crc8.h"
+#include "ad4170.h"
+
+NO_OS_DECLARE_CRC8_TABLE(ad4170_crc8);
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:50]: `static`

File: drivers/adc/ad4170/ad4170.c [hunk 5]
```diff
@@ -0,0 +1,2123 @@
...
+/* Swap bytes in a buffer with a given step
+ * Swap with step of 2:
+ * 	AA BB CC DD EE FF 00 11 becomes
+ *	BB AA DD CC FF EE 11 00
+ * Swap with step of 3:
+ * 	AA BB CC DD EE FF 00 11 22 becomes
+ * 	CC BB AA FF EE DD 22 11 00
+ * etc.
+ */
+static void _memswap64(void * buf, uint32_t bytes, size_t step)
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:185]: In the meantime, this function and `_is_bit_endian()` has been added to `no_os_util.c`, so we should use those instead.

File: drivers/adc/ad4170/ad4170.c [hunk 6]
```diff
@@ -0,0 +1,2123 @@
...
+			return -EBADMSG;
+	}
+
+	*reg_data = 0;
+	if (dev->big_endian)
+		memcpy((uint8_t *)(reg_data + 1) - reg_size, &buf[di], reg_size);
+	else {
+		memcpy(reg_data, &buf[di], reg_size);
+		_memswap64(reg_data, reg_size, reg_size);
+	}
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:301]: Can be replaced by `*reg_data = no_os_get_unaligned_be32()`? The endianess shouldn't matter in that case.

File: drivers/adc/ad4170/ad4170.c [hunk 7]
```diff
@@ -0,0 +1,2123 @@
...
+	 * Note: CS must be low for entire sequence. */
+	uint8_t buf[] = {0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
+			 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe,
+			 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xfe
+			};
+
+	if (!dev)
+		return -EINVAL;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, sizeof(buf));
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:424]: return directly

File: drivers/adc/ad4170/ad4170.c [hunk 8]
```diff
@@ -0,0 +1,2123 @@
...
+	int32_t ret;
+
+	if (!data)
+		return -EINVAL;
+
+	ret = ad4170_spi_reg_read(dev, AD4170_REG_DATA_16b, &reg);
+	if (ret < 0)
+		return ret;
+
+	*data = (uint16_t)reg;
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:449]: the cast is not needed

File: drivers/adc/ad4170/ad4170.c [hunk 9]
```diff
@@ -0,0 +1,2123 @@
...
+	int32_t ret;
+
+	if (!data || !status)
+		return -EINVAL;
+
+	ret = ad4170_spi_reg_read(dev, AD4170_REG_DATA_16b_STATUS, &reg);
+	if (ret < 0)
+		return ret;
+
+	*status = (uint8_t)reg;
```
Comment (CiprianRegus) [inline drivers/adc/ad4170/ad4170.c:474]: In cases like these, I think we usually use `no_os_field_get()` along with some macros for the bit mask of the register field. I'm fine with either option.

File: drivers/adc/ad4170/ad4170.c (hunk 4 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c:50]: it is static: https://github.com/analogdevicesinc/no-OS/blob/main/include/no_os_crc8.h#L48

File: drivers/adc/ad4170/ad4170.c (hunk 2 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c:737]: will do

File: drivers/adc/ad4170/ad4170.h (hunk 1 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.h:1066]: will do

File: drivers/adc/ad4170/ad4170.c (hunk 6 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c:301]: From memory, CPU endianess could be little or big, complicated by the fact that the data can come LSB or MSB first on the digital interface and this incoming data may not be aligned to powers of two. memswap64 has a step parameter which could be 3, as in the example. So I'm not touching this part, I'm pretty sure it's not possible to rewrite it using those macros.

File: drivers/adc/ad4170/ad4170.c (hunk 7 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c:424]: will do

File: drivers/adc/ad4170/ad4170.c (hunk 8 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c:449]: explicit casts carry meaning so i'll leave it as-is

**Comment by buha** [comment]
V3:
- addressed some of the points of the reviews above

Coming up with an excuse here for not implementing complex changes - this driver was reviewed in 2021 and is known to work so I'll be implementing only trivial changes.

**Comment by buha** [comment]
V4:
- reverted change of ad4170_config_reset to `static`, turns out it's prepended with `extern` in .h, I guess to publish the default device driver state after calling the reset API.

**Comment by buha** [comment]
V5:
- astyle

File: drivers/adc/ad4170/ad4170.h [hunk 10]
```diff
@@ -0,0 +1,1142 @@
...
+
+/* AD4170_REG_INTERFACE_CONFIG_B */
+#define AD4170_INTERFACE_CONFIG_B_SINGLE_INST_MSK		NO_OS_BIT(7)
+#define AD4170_INTERFACE_CONFIG_B_SHORT_INSTRUCTION_MSK		NO_OS_BIT(3)
+#define AD4170_INTERFACE_CONFIG_B_SHORT_INSTRUCTION(x)		(((x) & 0x1) << 3)
+
+/* AD4170_REG_INTERFACE_CONFIG_C */
+#define AD4170_INTERFACE_CONFIG_C_CRC_MSK			(NO_OS_BIT(7) | NO_OS_BIT(6) | NO_OS_BIT(1) | NO_OS_BIT(0))
+#define AD4170_INTERFACE_CONFIG_C_CRC(x)			(((~x) & 0x3) | (((x) << 6) & 0xC0))
+#define AD4170_INTERFACE_CONFIG_C_STRICT_REG_ACCESS_MSK	NO_OS_BIT(5)
```
Comment (danmois) [inline drivers/adc/ad4170/ad4170.h:138]: Pay attention to white spaces and alignment.

File: drivers/adc/ad4170/ad4170.h [hunk 11]
```diff
@@ -0,0 +1,1142 @@
...
+	/** Enable Continuous Transmit. This enables continuous transmit of the ADC Data register over TDM. */
+	AD4170_CONT_TRANSMIT_ON
+};
+
+/**
+ * @enum ad4170_mode
+ * @brief ADC Operating Mode.
+ */
+enum ad4170_mode {
+	/** Continuous Conversion Mode. ADC converts continuously on the enabled channel(s) using Sinc-based filters. */
```
Comment (danmois) [inline drivers/adc/ad4170/ad4170.h:455]: Should these comments be split on multiple lines?

Comment (buha) [reply drivers/adc/ad4170/ad4170.h:455]: hmmm, astyle didnt complain so i guess it's ok

File: drivers/adc/ad4170/ad4170.h [hunk 12]
```diff
@@ -0,0 +1,1142 @@
+/***************************************************************************//**
+ *   @file   ad4170.h
+ *   @brief  Header file for the ad4170 driver.
+ *   @author Darius Berghe (darius.berghe@analog.com)
+********************************************************************************
+ * Copyright 2020, 2023(c) Analog Devices, Inc.
```
Comment (danmois) [inline drivers/adc/ad4170/ad4170.h:6]: 2024?

File: drivers/adc/ad4170/ad4170.c [hunk 13]
```diff
@@ -0,0 +1,2119 @@
+/***************************************************************************//**
+ *   @file   ad4170.c
+ *   @brief  Source file for the ad4170 driver.
+ *   @author Darius Berghe (darius.berghe@analog.com)
+ *   @author Janani Sunil (janani.sunil@analog.com)
+ *   @author Mahesh Phalke (mahesh.phalke@analog.com)
+********************************************************************************
+ * Copyright 2020, 2023(c) Analog Devices, Inc.
```
Comment (danmois) [inline drivers/adc/ad4170/ad4170.c:8]: 2024?

File: drivers/adc/ad4170/ad4170.h (hunk 10 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.h]: will do

File: drivers/adc/ad4170/ad4170.h (hunk 12 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.h:6]: these dates are correct from the point of view of when development was done, work is being published only now...  no idea which one should we pick

**Comment by buha** [comment]
V6:
- fixed macro alignment

File: drivers/adc/ad4170/ad4170.c [hunk 14]
```diff
@@ -0,0 +1,2119 @@
...
+	}
+};
+
+static inline bool _is_addr_valid(struct ad4170_dev *dev, uint32_t reg_addr)
+{
+	uint8_t reg_size = AD4170_TRANSF_LEN(reg_addr);
+	uint32_t addr = AD4170_ADDR(reg_addr);
+
+	/* max readable address in short instruction mode */
+	if (dev->spi_settings.short_instruction && addr > 0x3F)
```
Comment (danmois) [inline drivers/adc/ad4170/ad4170.c:155]: Should we use macros for values like this one?

File: drivers/adc/ad4170/ad4170.c [hunk 15]
```diff
@@ -0,0 +1,2119 @@
...
+		return -EINVAL;
+
+	ret = ad4170_spi_reg_write_mask(dev,
+					AD4170_REG_INTERFACE_CONFIG_A,
+					AD4170_SW_RESET_MSK | AD4170_SW_RESETX_MSK,
+					AD4170_SW_RESET_MSK | AD4170_SW_RESETX_MSK);
+	if (ret)
+		return ret;
+
+	/* TODO: set proper value, it's currently undefined in datasheet. */
```
Comment (danmois) [inline drivers/adc/ad4170/ad4170.c:994]: Is there any datasheet available now?

Comment (buha) [reply drivers/adc/ad4170/ad4170.c:994]: good point, will try to find the right value

Comment (buha) [reply drivers/adc/ad4170/ad4170.c:994]: found it in the datasheet as Turn-On Settling Time, specified to 200us, so I will update accordingly

File: drivers/adc/ad4170/ad4170.c (hunk 14 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c]: i'd leave it as-is because it has something to do with a certain spi standard and we don't have headers for that

**Comment by jansunil** [comment]
Hi All- I've tested the latest version here and things look ok.

**Review by CiprianRegus — APPROVED** [review]

File: drivers/adc/ad4170/ad4170.h [hunk 16]
```diff
@@ -0,0 +1,1142 @@
...
+#define AD4170_REG_DATA_32b			(AD4170_R4B | 0x24)
+#define AD4170_REG_DATA_PER_CHANNEL(ch)		(AD4170_R4B | (0x28 + 4 * (ch)))
+#define AD4170_REG_PIN_MUXING			(AD4170_R2B | 0x68)
+#define AD4170_REG_CLOCK_CTRL			(AD4170_R2B | 0x6a)
+#define AD4170_REG_STANDBY_CTRL			(AD4170_R2B | 0x6c)
+#define AD4170_REG_POWER_DOWN_SW		(AD4170_R2B | 0x6e)
+#define AD4170_REG_ADC_CTRL			(AD4170_R2B | 0x70)
+#define AD4170_REG_ERROR_EN			(AD4170_R2B | 0x72)
+#define AD4170_REG_ERROR			(AD4170_R2B | 0x74)
+#define AD4170_REG_INFO1			(AD4170_R2B | 0x76) /* TODO: implement this when it's specified in doc. */
```
Comment (jansunil) [inline drivers/adc/ad4170/ad4170.h:92]: Should we remove the TODO before release?

File: drivers/adc/ad4170/ad4170.c [hunk 17]
```diff
@@ -0,0 +1,2119 @@
...
+		no_os_gpio_get_value(dev->gpio_dig_aux1, &rdyb);
+	};
+	if (!timeout)
+		return -ETIMEDOUT;
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, sz);
+	if (ret)
+		return ret;
+
+	// TODO: handle 32b width. By default 24b data width is selected.
```
Comment (jansunil) [inline drivers/adc/ad4170/ad4170.c:781]: Should we remove the TODOs?

File: drivers/adc/ad4170/ad4170.h (hunk 16 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.h:92]: good point, turns out INFO1 is not in the latest datasheet, and is also unused in the driver so I'm removing it

File: drivers/adc/ad4170/ad4170.c (hunk 17 above)
Comment (buha) [reply drivers/adc/ad4170/ad4170.c:781]: I'm not sure about this one, looks like there is some missing functionality but I can't add it without hardware setup, so I'd rather leave it in for now.

Comment (buha) [reply drivers/adc/ad4170/ad4170.c:781]: actually i re-read the datasheet and looks like continuous read works with 24b mode only so I guess we're good, I'm removing the TODO and the 32b handling that was commented out

**Comment by buha** [comment]
V7:
- removed TODOs
- removed non existing register INFO1

**Comment by buha** [comment]
I see it has been approved and tested, I will merge now

**Review by danmois — APPROVED** [review]
Looks good.

File: drivers/adc/ad4170/ad4170.h [hunk 18]
```diff
@@ -0,0 +1,1141 @@
...
+int ad4170_set_dac_inputb(struct ad4170_dev *dev, uint16_t code);
+int ad4170_dac_sw_ldac(struct ad4170_dev *dev, bool polarity);
+int ad4170_dac_sw_toggle(struct ad4170_dev *dev, bool polarity);
+int ad4170_dac_hw_toggle(struct ad4170_dev *dev, bool polarity);
+int ad4170_init(struct ad4170_dev **device,
+		struct ad4170_init_param *init_param);
+int ad4170_remove(struct ad4170_dev *dev);
+int ad4170_regmap(struct ad4170_dev *dev);
+#endif
+
```
Comment (amiclaus) [inline drivers/adc/ad4170/ad4170.h:1141]: nit: looks like multiple blank lines

File: drivers/adc/ad4170/ad4170.c [hunk 19]
```diff
@@ -0,0 +1,2116 @@
...
+			printf("error %d while reading reg 0x%x", ret, AD4170_ADDR(regs[ri]));
+			return ret;
+		}
+
+		printf("0x%.4x: 0x%x\n", AD4170_ADDR(regs[ri]), reg);
+	}
+
+	return 0;
+}
+
```
Comment (amiclaus) [inline drivers/adc/ad4170/ad4170.c:2116]: ditto

File: drivers/adc/ad4170/ad4170.c [hunk 20]
```diff
@@ -0,0 +1,2116 @@
...
+		AD4170_REG_DAC_HW_LDAC_MASK,
+		AD4170_REG_DAC_DATA(0),
+		AD4170_REG_DAC_SW_TOGGLE_TRIGGERS,
+		AD4170_REG_DAC_SW_LDAC_TRIGGERS,
+		AD4170_REG_DAC_INPUTA(0),
+		AD4170_REG_DAC_INPUTB(0),
+	};
+
+	uint32_t ri;
+	printf("[register]: [value]\n");
```
Comment (amiclaus) [inline drivers/adc/ad4170/ad4170.c:2103]: let's go for `pr_info` (or even `pr_debug` in this particular case). same applies for all the other `printf` throughout the code.

**Review by amiclaus — COMMENTED** [review]
some minor comments on my side.

**Pull request merged**